swfiles/sw_fitpowder.m

```diff
@@ -0,0 +1,556 @@
+            obj.y = obj.y(:, ikeep);
+            obj.e = obj.e(:, ikeep);
+        end
+
+        function bg = calc_background(obj, params)
+            % add background
+            istart = obj.nparams_model + 1;
+            if obj.background_strategy == "planar"
+                bg_params = num2cell(params(istart:istart+obj.nparams_bg-1));
+                bg = obj.fbg(obj.ebin_cens, obj.modQ_cens, bg_params{:});
+            elseif obj.ndim == 1
+                % add energy dependent background to each cut
+                bg = zeros(size(obj.y));
+                for icut = 1:obj.ncuts
+                    iend = istart + +obj.nparams_bg - 1;
+                    bg_params = num2cell(params(istart:iend));
+                    bg(:,icut) = obj.fbg(obj.ebin_cens(:), bg_params{:});
+                    istart = iend + 1;
+                end
+            end
+        end
+
+        function set_caching(obj, do_cache)
+            obj.do_cache = do_cache;
+            if ~obj.do_cache
+                obj.clear_cache()
+            end
+        end
+
+        function clear_cache(obj)
+            obj.ycalc_cached = [];
+            obj.model_params_cached = [];
+        end
+
+        function [ycalc, bg] = calc_spinwave_spec(obj, params)
+            model_params = reshape(params(1:obj.nparams_model), 1, []);
+            if obj.do_cache && ~isempty(obj.model_params_cached) && all(abs(model_params - obj.model_params_cached) < 1e-10)
+                ycalc = obj.ycalc_cached;  % do not repeat calc
+            else
+                % set spinw model interactions
+                % pass params as row-vector as expected by matparser
+                obj.fit_func(obj.swobj, model_params);
+                % simulate powder spectrum
+                spec = obj.swobj.powspec(obj.modQ_cens, obj.powspec_args);
+                spec = sw_instrument(spec, obj.sw_instrument_args);
+                ycalc = spec.swConv;
+                % cache if required
+                if obj.do_cache
+                    obj.ycalc_cached = spec.swConv;
+                    obj.model_params_cached = model_params;
+                end
+            end
+            % scale
+            ycalc = params(end)*ycalc;
+            % calc background
+            bg = calc_background(obj, params);
+            if obj.background_strategy == "planar"
+                ycalc = ycalc + bg;  % add planar bg before any rebinning
+                if obj.ndim == 1
+                    % integrate nQ |Q| points for each cut
+                    ycalc = obj.rebin_powspec_to_1D_cuts(ycalc);
+                end
+            else
+                % integrate nQ |Q| points for each cut
+                ycalc = obj.rebin_powspec_to_1D_cuts(ycalc);
+                % add background to individual cuts after rebin
+                ycalc = ycalc + bg;
+            end
+
+        end
+
+        function resid_sq_sum = calc_cost_func(obj, params)
+            % evaluate fit function
```

Comment:
Thanks for this code - have done this in commit https://github.com/SpinW/spinw/pull/174/commits/feb6cb3d37df7edbdfb4fbddafd1085cbda5c819 - but it plots each function evaluation, so e.g. in `lm3` every time it needs to compute finite-differences - I think to do this properly we'd need to send some signal from the minimiser, or call the minimiser in a loop with 1 iteration at a time (very slow...)

Reply:
Thanks. Yeah, it's not ideal but a more efficient (faster!) approach will need some knowledge of the minimizer and so would not be general... 

In my [original code](https://github.com/SpinW/spinw/blob/d78b10e3c62db51fa6b225d03a0f2d8232df4eb0/swfiles/%40spinw/fitpow.m#L654-L658) I had a persistent variable `lines` which are just the calculated lines. Then at each iteration instead of calling `clf` I would delete the lines of the previous iteration and plot new lines - so the axes labels, titles and the data points would stay the same, only the simulated lines would be replotted. 

This is a bit faster but to be honest, I just profiled a run for a medium-sized system (16 spins) and the code spent 40s in total with the plotting (17s for `plot_1d_or_2d` and 23s in `drawnow`) for 246 evaluations (so 0.16s per evaluation), by contrast it spent 400s (1.6s per evaluation) running `calc_spinwave_spec` (almost all of that time spent in the `swloop` mex file) - so the plotting adds an addition 10% running time... (of course this fraction will be higher for smaller systems with fewer spins).

I think possibly the easiest thing to do is to change `liveplot` from a bool to an integer (maybe called `liveplot_interval`) where only every `liveplot_interval`th evaluation is plotted (e.g. set `liveplot_interval=10` will plot every 10th evaluation and thus reduce the plotting overhead for medium systems to 1% of the total running time), with a default of `0` (no live plots - I know it's not how ratios work... but 🤷 )